File: VB/B138492/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraBars
Imports System.Collections
Imports DevExpress.XtraTreeList.ViewInfo

Namespace B138492
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
			treeList1.ExpandAll()
		End Sub

		Private Sub OnAddNodeBarButtonItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles bbiAddNode.ItemClick
			Dim treeList As TreeList = TryCast(ActiveControl, TreeList)
			If treeList Is Nothing OrElse treeList.FocusedNode Is Nothing Then
				Return
			End If
			AppendEmptyNode(treeList, treeList.FocusedNode.ParentNode)
		End Sub

		Private Sub OnAddChildBarButtonItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles bbiAddChild.ItemClick
			Dim treeList As TreeList = TryCast(ActiveControl, TreeList)
			If treeList Is Nothing OrElse treeList.FocusedNode Is Nothing Then
				Return
			End If
			AppendEmptyNode(treeList, treeList.FocusedNode)
		End Sub

		Private Sub OnDeleteNodeBarButtonItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles bbiDeleteNode.ItemClick
			Dim treeList As TreeList = TryCast(ActiveControl, TreeList)
			If treeList Is Nothing OrElse treeList.FocusedNode Is Nothing Then
				Return
			End If
			If treeList.Nodes.Count = 1 AndAlso treeList.FocusedNode.ParentNode Is Nothing Then
				Return
			End If
			treeList.DeleteNode(treeList.FocusedNode)
		End Sub

		Private Sub AppendEmptyNode(ByVal treeList As TreeList, ByVal parent As TreeListNode)
			Dim nodeData As New ArrayList()
			For i As Integer = 0 To treeList.Columns.Count - 1
				nodeData.Add(Nothing)
			Next i
			treeList.AppendNode(nodeData.ToArray(), parent)
		End Sub

		Private Sub OnShowTreeListMenuClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles bbiShowTreeListMenu.ItemClick
			Dim treeList As TreeList = TryCast(ActiveControl, TreeList)
			If treeList Is Nothing OrElse treeList.FocusedNode Is Nothing Then
				Return
			End If
			Dim focusedNodeBounds As Rectangle = (CType(treeList.ViewInfo.RowsInfo(treeList.FocusedNode).Cells(0), CellInfo)).Bounds
			Dim menuLocation As New Point(focusedNodeBounds.Left + 3, focusedNodeBounds.Bottom + 1)
			menuLocation = treeList.PointToScreen(menuLocation)
			treeListMenu.ShowPopup(menuLocation)
		End Sub
	End Class
End Namespace

File: CS/B138492/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraBars;
using System.Collections;
using DevExpress.XtraTreeList.ViewInfo;

namespace B138492 {
    public partial class Form1 :Form {
        public Form1() {
            InitializeComponent();
            treeList1.ExpandAll();
        }

        private void OnAddNodeBarButtonItemClick(object sender, ItemClickEventArgs e) {
            TreeList treeList = ActiveControl as TreeList;
            if (treeList == null || treeList.FocusedNode == null) return;
            AppendEmptyNode(treeList, treeList.FocusedNode.ParentNode);
        }

        private void OnAddChildBarButtonItemClick(object sender, ItemClickEventArgs e) {
            TreeList treeList = ActiveControl as TreeList;
            if (treeList == null || treeList.FocusedNode == null) return;
            AppendEmptyNode(treeList, treeList.FocusedNode);
        }

        private void OnDeleteNodeBarButtonItemClick(object sender, ItemClickEventArgs e) {
            TreeList treeList = ActiveControl as TreeList;
            if (treeList == null || treeList.FocusedNode == null) return;
            if (treeList.Nodes.Count == 1 && treeList.FocusedNode.ParentNode == null) return;
            treeList.DeleteNode(treeList.FocusedNode);
        }

        private void AppendEmptyNode(TreeList treeList, TreeListNode parent) {
            ArrayList nodeData = new ArrayList();
            for (int i = 0; i < treeList.Columns.Count; i++)
                nodeData.Add(null);
            treeList.AppendNode(nodeData.ToArray(), parent);
        }

        private void OnShowTreeListMenuClick(object sender, ItemClickEventArgs e) {
            TreeList treeList = ActiveControl as TreeList;
            if (treeList == null || treeList.FocusedNode == null) return;
            Rectangle focusedNodeBounds = ((CellInfo)treeList.ViewInfo.RowsInfo[treeList.FocusedNode].Cells[0]).Bounds;
            Point menuLocation = new Point(focusedNodeBounds.Left + 3, focusedNodeBounds.Bottom + 1);
            menuLocation = treeList.PointToScreen(menuLocation);
            treeListMenu.ShowPopup(menuLocation);
        }
    }
}